### PR TITLE
Support shutdown hooks with signals

### PR DESCRIPTION
**Pull request by ekrich**
Now you can use Ctrl-C or `kill -s SIGTERM <pid>` and shutdown hooks will be called and the process will exit gracefully.

Here is the `sandbox` I used - not sure how to test.
```scala
object Test {
  def main(args: Array[String]): Unit = {
    println("Shutdown test - use Ctrl-C or send SIGTERM")
    // Runtime
    //   .getRuntime()
    //   .addShutdownHook(new Thread() {
    //     override def run(): Unit = messageThread()
    //   })
    // or this
    sys.addShutdownHook(messageThread())
    while (true) {
      try Thread.sleep(1000)
      catch {
        case e: InterruptedException => {
          println("Interrupted")
        }
      }
    }
  }
  def messageThread() = (0 to 5).foreach { i =>
    println("Shutting down")
    Thread.sleep(500)
  }
}
```

### CODE REVIEW

File: javalib/src/main/scala/java/lang/Runtime.scala [hunk 1]
```diff
@@ -67,7 +79,7 @@ class Runtime private () {
       .asInstanceOf[Array[Thread]]
       .sorted(Ordering.by[Thread, Int](-_.getPriority()))
       .foreach { t =>
-        try t.run()
+        try t.start()
```
Comment (WojciechMazur) [inline javalib/src/main/scala/java/lang/Runtime.scala:82]: We should revert this change, `runHooksSequential` is run only in the singlethreaded mode

File: javalib/src/main/scala/java/lang/Runtime.scala [hunk 2]
```diff
@@ -81,6 +93,7 @@ class Runtime private () {
       if (isMultithreadingEnabled) runHooksConcurrent()
       else runHooksSequential()
     }
+    exit(0)
```
Comment (WojciechMazur) [inline javalib/src/main/scala/java/lang/Runtime.scala:96]: What if the shutdown was started by unhandled thread or signal? In such case we would report successful exit during program failure

**Review by WojciechMazur — CHANGES_REQUESTED** [review]
Overall LGTM, only 2 things I don't think are correct

File: javalib/src/main/scala/java/lang/Runtime.scala (hunk 1 above)
Comment (ekrich) [reply javalib/src/main/scala/java/lang/Runtime.scala:82]: Ok, I didn't understand that one because you normally call `start()` on a thread.

File: javalib/src/main/scala/java/lang/Runtime.scala (hunk 2 above)
Comment (ekrich) [reply javalib/src/main/scala/java/lang/Runtime.scala:96]: This one I put in because it was hanging after the shutdown hooks ran and I had no idea how to solve that.

**Comment by ekrich** [comment]
This seems to work now. Avoid race without calling shutdown hooks twice and call `exit` after handling shutdown hooks. Not sure if this is 100% correct but it seems ok.

**Comment by ekrich** [comment]
I'll need to fix this but not sure what value to put.
```
[error] Test scala.scalanative.linker.MinimalRequiredSymbolsTest.multithreading failed: java.lang.AssertionError: 
[error] Found more symbols then expected, config={debugMetadata=false, multithreading=true, targetTriple=x86_64-unknown-unknown}:
[error] Expected at most: {types=1050, members=6600, total=7650}
[error] Found:            {types=1028, members=6640, total=7668}
[error] Diff:             {types=-22, members=40, total=18}
```

File: javalib/src/main/scala/java/lang/Runtime.scala [hunk 3]
```diff
@@ -18,6 +19,17 @@ class Runtime private () {
...
     stdlib.atexit(() => Runtime.getRuntime().runHooks())
   }
+
+  lazy val setupSignalHandler = {
+    signal.signal(signal.SIGINT, handleSignal(_))
+    signal.signal(signal.SIGTERM, handleSignal(_))
+  }
+
+  private def handleSignal(sig: CInt): Unit = {
+    Runtime.getRuntime().runHooks()
```
Comment (WojciechMazur) [inline javalib/src/main/scala/java/lang/Runtime.scala:29]: I think it won't work when we'll have multiple threads running. Starting `runHooks` would also start `scala.scanative.runtime.JoinNonDaemonThreads` which would wait for all thread to finish. We should at least remove this shutdown hook. 
Also signal handler might be run by any of the threads. Which might affect its execution.

**Review by WojciechMazur — CHANGES_REQUESTED** [review]
For the tools tests it's fine to just update the values to the number of found symbols/classes + alignment to `% 50` so small changes won't affect it much. Maybe we should remove this test. 

Also, this use cases seems to be good candidate for scripted test. We probably should be able to create a custom sbt task to start compiled program (hopefully using multitple threads running in the lop) and create a task to run `Process.destroy` - I guess it should be sending the SIGINT / SIGTERM signal

**Comment by catap** [comment]
Can it help with https://github.com/scala-native/scala-native/issues/2483 ?

File: javalib/src/main/scala/java/lang/Runtime.scala [hunk 4]
```diff
@@ -18,6 +19,17 @@ class Runtime private () {
   lazy val setupAtExitHandler = {
     stdlib.atexit(() => Runtime.getRuntime().runHooks())
   }
+
+  lazy val setupSignalHandler = {
+    signal.signal(signal.SIGINT, handleSignal(_))
+    signal.signal(signal.SIGTERM, handleSignal(_))
+  }
```
Comment (catap) [inline javalib/src/main/scala/java/lang/Runtime.scala:26]:
```suggestion
  lazy val setupSignalHandler = {
    signal.signal(signal.SIGHUP, handleSignal(_))
    signal.signal(signal.SIGINT, handleSignal(_))
    signal.signal(signal.SIGQUIT, handleSignal(_))
    signal.signal(signal.SIGTERM, handleSignal(_))
  }
```
let's add a few more expected reason to quit

Comment (ekrich) [reply javalib/src/main/scala/java/lang/Runtime.scala:26]: As far as I understand "The JVM runs shutdown hooks only in case of normal terminations." - what this seems to mean is Ctrl-C `SIGINT` and `SIGTERM`. This is also in the current Test comments.

Comment (catap) [reply javalib/src/main/scala/java/lang/Runtime.scala:26]: > The JVM catches signals to implement shutdown hooks for unexpected termination. The JVM uses SIGHUP, SIGINT, and SIGTERM to initiate the running of shutdown hooks.
> Applications embedding the JVM frequently need to trap signals such as SIGINT or SIGTERM, which can lead to interference with the JVM signal handlers. The -Xrs option is available to address this issue. When -Xrs is used, the signal masks for SIGINT, SIGTERM, SIGHUP, and SIGQUIT aren't changed by the JVM, and signal handlers for these signals aren't installed.

https://docs.oracle.com/en/java/javase/21/docs/specs/man/java.html

Comment (ekrich) [reply javalib/src/main/scala/java/lang/Runtime.scala:26]: Ok, I'll add `SIGHUP`.

Comment (ekrich) [reply javalib/src/main/scala/java/lang/Runtime.scala:26]: On second thought, we need to try and get this to work properly this way first as we are using `clib` and `SIGHUP` is not a member. I second pass could use `posixlib` and something special for Windows to make it JVM compliant.

Comment (catap) [reply javalib/src/main/scala/java/lang/Runtime.scala:26]: Keep in mind that `SIGHUP` is a signal which is send when SSH connection is hand up due network issue.

Without it, shutdown hook at CLI utilite won't work on this case and it's show stoper for real life use case I guess.

Comment (ekrich) [reply javalib/src/main/scala/java/lang/Runtime.scala:26]: No problem but we have nothing now and to make it cross platform is more complex so it might be better to have something now.

**Review by WojciechMazur — APPROVED** [review]
LGTM for the single threaded execution, I have prepared a variant handling the multithreaded case including handling on the GC side. I'll add the follow up PR soon

**Comment by ekrich** [comment]
Thanks, I was looking into the multi-threaded version and it was quite a bit more involved.
